gh-109599: Expose `CapsuleType` via the `_types` module

AA-Turner opened the pull request:
Micro-optimisation, but we should avoid re-calculating `CapsuleType` every time it is accessed in the `types` module.

A

<!-- gh-issue-number: gh-109599 -->
* Issue: gh-109599
<!-- /gh-issue-number -->

AA-Turner committed:
Cache ``types.CapsuleType`` on first compution

AA-Turner committed:
Add a _types module exposing `CapsuleType` and `SimpleNamespace`

AA-Turner committed:
Use the `_types` module in `types`

ZeroIntensity commented on Modules/_typesmodule.c:
```diff
@@ -0,0 +1,37 @@
+static int
+_types_exec(PyObject *m)
+{
+    if (PyModule_AddObjectRef(m, "CapsuleType", (PyObject *)&PyCapsule_Type) < 0) {
```
There's [a lot](https://github.com/ZeroIntensity/cpython/blob/7bd08c8e8a18e5bfeb6ed386050fb1ed04486e2a/Modules/_typesmodule.c#L1265) of potential here; why expose only `CapsuleType` and `SimpleNamespace`?

Edit: Oh, I didn't see your other comment.

AA-Turner replied:
`CapsuleType` as there is no way of getting it save via the extension module, and `SimpleNamespace` to remove the remaining use of `sys`.

Remember that the stdlib is often copied and reused by other implementations: by moving everything to C, we gain performance, but make things harder for MicroPython, PyPy, RustPython, etc.

ZeroIntensity replied:
I'm not sure we actually gain much performance here; `type()` and everything else that `types` uses have gotten pretty fast, and switching things over to an extension didn't give huge speedups from when I benchmarked it.

That said, it's actually much easier for us to add types from the extension module. As long as we keep the pure-Python implementation as a fallback, I think it's worth adding the types that are directly exposed in C to the accelerator.

AA-Turner replied:
Compare `python -Ximporttime -Sc "from types import CapsuleType"`. The benefit is from avoiding `_socket`. I agree replacing the definitions of the rest of `types.py` would be far more marginal, though.

AA-Turner replied:
I've opened #132103 as a potential followup.

AA-Turner committed:
Apply configure changes

AA-Turner committed:
fixup! Apply configure changes

AA-Turner committed:
fixup! Apply configure changes

AA-Turner committed:
Remove `internal/` prefix

zooba commented on Lib/types.py:
```diff
@@ -331,11 +333,4 @@ def wrapped(*args, **kwargs):
-
-__all__ = [n for n in globals() if n[:1] != '_']
-__all__ += ['CapsuleType']
+__all__ = [n for n in globals() if not n.startswith('_')]
```
This is the default behaviour of `__all__`, isn't it? So we shouldn't need to specify it.

AA-Turner replied:
Good point, removed.

picnixz replied:
But could it have been used by existing packages that want to extend `types` maybe? or use `if x in types.__all__`? is `__all__` automatically defined in this case?

AA-Turner replied:
`__all__` is only used in the data model for star imports: https://docs.python.org/3/reference/simple_stmts.html#the-import-statement. `dir()` should be used for getting the names defined in a module.

picnixz replied:
Yes, but what if someone overrides the `types.py` module and re-export the same names?

```py
import types
from types import *
__all__ = []
__all__.extend(types.__all__)
__all__.extend(['MyType'])

class MyType:
   pass
```

But if you think it's not worth to keep this kind of possible constructions, I'm fine with removing `__all__`.

pitrou replied:
> Yes, but what if someone overrides the `types.py` module and re-export the same names?

I don't think that's a supported use of stdlib modules, so they are on their own.

AA-Turner replied:
I've reinstated `__all__`, as without it `pydoc` fails to pick up several members of the module, and added an explanatory comment. See also https://github.com/python/cpython/issues/67692, when it was first added.

zooba reviewed (commented):
I like this approach, though would want someone else's signoff because I'm not as familiar with the `Modules/Setup` changes. Provided the module is always statically linked, it ought to be fine.

AA-Turner committed:
Remove ``types.__all__``

`from types import *` will ignore underscore-prefixed names by default, so defining `__all__` doesn't add anything.

AA-Turner committed:
Reinstate __all__ for pydoc

XuehaiPan commented on Lib/types.py:
```diff
@@ -60,7 +60,9 @@ def _m(self): pass
 MemberDescriptorType = type(FunctionType.__globals__)
 
-del sys, _f, _g, _C, _c, _ag, _cell_factory  # Not for export
+CapsuleType = _types.CapsuleType
```
A nit suggestion: `mypy` has better support for re-export imports than type alias assignment:

```python
from _types import CapsuleType as CapsuleType
# vs.
CapsuleType: type = _types.CapsuleType
```

AA-Turner replied:
From mypy's perspective the entire stdlib is overwritten by typeshed.

A

ZeroIntensity reviewed (commented):
Should we add pure-Python fallbacks for when `_types` isn't available?

On the one hand, it's good for other implementations that want to provide the `types` module, but on the other hand, `types` is pretty CPython-specific anyway; it would probably need impl-specific changes, even if it's pure-Python.

LGTM otherwise.

ZeroIntensity reviewed (approved):
Sounds good. The implementation looks fine, and unlike most C accelerators, I think `_types` can actually *simplify* some of the code, because we don't have to use silly tricks to get the types at runtime.

Thanks for doing this!
